### DISCUSSION TITLE
How to create a KeycloakSession in a provider on the server-side

### DISCUSSION THREAD

**Discussion by blevine** [keycloak/keycloak#34787] [Q&A]
Most of the Keycloak APIs I need for the [Keycloak GraphQL API](https://github.com/blevine/keycloak-graphql/) I'm building require a KeycloakSession. Luckily, the Keycloak REST infrastructure creates one (a QuarkusKeycloakSession) in the REST provider on each REST handler invocation. For subscriptions, the query is re-executed every time the subscription fires and the resolvers need to be passed a KeycloakSession. So the question is: **how do we create a valid KeycloakSession on the server side whenever the subscription event is fired**? This session should represent the User/Principal that made the original request to create the subscription. I can probably associate information about that User with the subscription.

**Comment by danielFesenmeyer** [comment]
What about KeycloakSessionFactory#createSession?
After you have the session, you have to make sure to set the correct information on #getContext (realm, client, authenticationSession).

**Reply by blevine** [reply]
Yes. In my case, I'm using QuarkusKeycloakSesionFactory.createSession(). The problem is setting up the context correctly. For example, there is no active client connection when the subscription event is fired to context.createClientConnection() fails. Trying to figure out how to set that correctly or stub it.

**Reply by shawkins** [reply]
@blevine if it's your code calling getConnection(), can't you just catch the exception and keep going? Or are you passing it to existing logic that will need some kind of dummy ClientConnection?
